### ISSUE TITLE
HELP WANTED

### ISSUE THREAD

**Issue by grezniczek** [grezniczek/redcap_em_config_study#6]
If anybody wants to help making this a reality - please chime in with starting a conversation or pull requests !

Thanks!

**Comment by tertek** [comment]
I would like to help. 
What is the current state?

**Comment by grezniczek** [comment]
@tertek - this is great! Thanks!

Well, nothing has happened since Mar 8, 2021 (last commit). Since quite a bit has changed in the EM Framework in more recent REDCap versions, this EM here doesn't even work currently. I'd have to check what broke. Normally, when installed, it would pop up it's own configuration interface whenever one navigates to the _External Modules - Module Manager_ page.

Maybe I can find some time in the next couple of days/weeks to get it working again.

Feel free to have a look yourself, if you want.

**Comment by tertek** [comment]
@grezniczek thanks for the state info.

As far as I can understand from the notice in the official external module docs included within REDCap, this configuration interface is aimed to be implemented into the EM Framework correct?

![image](https://user-images.githubusercontent.com/75415872/166630401-18cf3162-6829-4461-ad7e-dd9acd4898af.png)

I will have a look into it as well and try to get it working.
Thanks!

**Comment by grezniczek** [comment]
@tertek , yes. This was the plan. If I remember correctly, the basic interface functionality was more or less complete. Most lacking are implementations of the various setting types, such as file uploads (I believe I was working on this when I stopped), project lists, event lists, etc. The idea here was to support dependencies (specified in config.json), e.g., to select a form/event pair, first select event, then limit form selector to forms on this event only.

But the first goal would be to have something that is fully compatible with what the EM Framework docs say should work (which is not true for the built-in settings dialog).

**Comment by grezniczek** [comment]
@tertek - I have fixed this now so that it will work again, even with some enhancements.

After installing it as a regular module, you'll get a `Configure+` button for each module on the module management pages (in projects only when this module is enabled in the respective project). Clicking this new button will load the enhanced configuration dialog for that module.

**Comment by tertek** [comment]
@grezniczek That is great! I am going to try it out as soon as I have finished up some of my waiting tasks. I will give you feedback here and then we can maybe plan how to proceed?

Looking forward on a good collab!
An improved EM interface is much needed IMO.

**Comment by AndrewPoppe** [comment]
@grezniczek I'm interested in helping out with this project. Is there something you'd recommend starting with? 

I'm seeing that the navigation tabs don't seem to be working in REDCap v13.10.05 but are working in v13.1.37. There are other differences between my two dev servers, so it might not be the RC version. However with the Bootstrap update that happened around 13.4/13.5 I would guess it is likely.

Anyway, let me know if you have some advice on how to get started.

**Comment by grezniczek** [comment]
@AndrewPoppe - I invited you as a collaborator to this project.

Yes, the different Bootstrap versions are likely the main issue. This project hasn't been touched in quite a while.

Currently, it seems that in BS5, all events are not correctly triggered. So the first step would be to adapt all event handling from BS4 to BS5-style. Some display issues will certainly have to be fixed as well.

**Comment by AndrewPoppe** [comment]
Thanks @grezniczek  - I just pushed some commits. If you'd rather I work on a separate branch and merge later, let me know. I hope the small changes I made just now are not a problem.

**Comment by grezniczek** [comment]
@AndrewPoppe -- No, this is great! Thanks!

**Comment by AndrewPoppe** [comment]
@grezniczek I've added some more fields. To do this, I included some more 3rd part libs:

- [Coloris](https://github.com/mdbassit/Coloris) - for colorpicker
- [Select2](https://select2.org/) - for enhanced select elements

I used TinyMCE for rich text (I think I'm safe assuming that will be loaded by REDCap).

I was thinking of using [flatpickr](https://flatpickr.js.org/) for date/datetime fields, since I've used it before.

There are obviously many other options for each of these, so I'm curious what your preference is.

**Comment by grezniczek** [comment]
@AndrewPoppe - Anything is fine, I guess. Of course, in case REDCap already brings a third party lib for some purpose, it would be preferable to use the "built-in" package (unless another would be much more beneficial).

REDCap will move to TinyMCE 6 soon, I believe.

For dates, I'd stick to jQuery.datepicker, because it's strings are covered by REDCap's .ini files and also - why not?

Edit: 
Oh, same for the colorpicker. Maybe use the one used by REDCap?
REDCap uses select2, so no issue there.

But if there are good reasons to not use the ones already shipped with REDCap, then I guess that would be fine, too.

**Comment by AndrewPoppe** [comment]
@grezniczek Okay that makes sense. I've changed the color picker to use the built-in spectrum color picker. One downside there is that it doesn't handle alpha very well, so I've disabled it. (Incidentally, in my testing, the color picker in the current EM Framework seems to support transparency, but it saves the value as a 6-digit hex since the preferred format is hex and the spectrum code does not handle 8-digit hex codes)

The other possible downside is that it does not seem to be actively maintained, but I suppose if that becomes a problem then we would just move to whatever REDCap moves to.

The same can be said for the time picker REDCap uses I suppose. The packaged time picker is [no longer actively maintained](https://github.com/trentrichardson/jQuery-Timepicker-Addon), but again REDCap can lead the way on choosing a new tool.

Let me know if you would like to discuss my additions further, otherwise I'll keep plugging along :)

**Comment by grezniczek** [comment]
@AndrewPoppe - Hmm, not longer being maintained (I wasn't aware of this) may be a good reason to move away from them - I'll leave that up to you.

Any contribution to this project is a positive one! So please keep plgging!
Thanks!

Eventually, this configuration editor (if it ever reaches feature parity with the built-in one) will hopefully replace the EM config editor. At least then, a general switch of REDCap as a whole (Core + EM Framework) off of some of the no longer maintained packages could/should be discussed.

**Comment by AndrewPoppe** [comment]
@grezniczek The spectrum color picker does not state explicitly that it is not actively maintained (like the time picker does), but it has had no activity since 2020 and minor activity leading up to that. But personally I think supporting both transparency and hex strings might be worth thinking about an alternative. But then again I don't know Coloris well and haven't vetted the code, so that's reason enough not to use it...

There's enough left to do without me getting too hung up trying to find the best libraries to use :)
